### PR TITLE
Gamma correction for dwrite fonts

### PR DESCRIPTION
**Pull request by changm**
Same as https://github.com/servo/webrender/pull/774#pullrequestreview-18222829 but for dwrite fonts!

---
This change is [Reviewable](https://reviewable.io/reviews/servo/webrender/775)

### CODE REVIEW

**Comment by glennw** [comment]
@vvuk Could you take a look over this?

File: webrender/src/platform/windows/font.rs [hunk 1]
```diff
@@ -17,8 +18,11 @@ lazy_static! {
     };
 }
 
+#[allow(dead_code)]
```
Comment (vvuk) [inline webrender/src/platform/windows/font.rs]: `dead_code` here, why?

File: webrender/src/platform/windows/font.rs [hunk 2]
```diff
@@ -60,14 +72,56 @@ impl FontContext {
...
+                let r = pixel[0];
+                let g = pixel[1];
+                let b = pixel[2];
+                print!("({}, {}, {}) ", r, g, b, );
+            }
+            println!("");
+        }
+    }
+
+    fn get_dwrite_texture_type(&self, render_mode: FontRenderMode) ->
```
Comment (vvuk) [inline webrender/src/platform/windows/font.rs]: This should be a free function, not part of this struct -- just `fn dwrite_texture_type` in this file.

File: webrender/src/platform/windows/font.rs [hunk 3]
```diff
@@ -60,14 +72,56 @@ impl FontContext {
...
+                               dwrote::DWRITE_TEXTURE_TYPE {
+        match render_mode {
+            FontRenderMode::Mono => dwrote::DWRITE_TEXTURE_ALIASED_1x1 ,
+            FontRenderMode::Alpha |
+            FontRenderMode::Subpixel => dwrote::DWRITE_TEXTURE_CLEARTYPE_3x1,
+        }
+    }
+
+    fn get_dwrite_measure_mode(&self, render_mode: FontRenderMode) ->
```
Comment (vvuk) [inline webrender/src/platform/windows/font.rs]: Also should be a free function

File: webrender/src/platform/windows/font.rs [hunk 4]
```diff
@@ -60,14 +72,56 @@ impl FontContext {
...
+                               dwrote::DWRITE_MEASURING_MODE {
+        match render_mode {
+            FontRenderMode::Mono => dwrote::DWRITE_MEASURING_MODE_GDI_NATURAL,
+            FontRenderMode::Alpha => dwrote::DWRITE_MEASURING_MODE_GDI_NATURAL,
+            FontRenderMode::Subpixel => dwrote::DWRITE_MEASURING_MODE_GDI_CLASSIC,
+        }
+    }
+
+    fn get_dwrite_render_mode(&self, render_mode: FontRenderMode) ->
```
Comment (vvuk) [inline webrender/src/platform/windows/font.rs]: Also should be a free function

File: webrender/src/platform/windows/font.rs [hunk 5]
```diff
@@ -85,98 +139,88 @@ impl FontContext {
...
+        let analysis = self.create_glyph_analysis(font_key, size, glyph, render_mode);
+
+        let texture_type = self.get_dwrite_texture_type(render_mode);
+        Some(self.get_glyph_dimensions_with_analysis(analysis, texture_type))
+    }
 
-        let pixels = analysis.create_alpha_texture(tex_type, bounds);
-        let rgba_pixels = match render_mode.unwrap() {
+    // DWRITE gives us values in RGB. WR expects it in RGBA. Note, CG returns in BGR
+    // TODO: probably should fix that
```
Comment (vvuk) [inline webrender/src/platform/windows/font.rs]: Fix what?

File: webrender/src/platform/windows/font.rs [hunk 6]
```diff
@@ -85,98 +139,88 @@ impl FontContext {
...
+                let mut rgba_pixels = vec![0 as u8; length * 4];
+                for i in 0..length {
                     // TODO(vlad): we likely need to do something smarter
-                    let alpha = (pixels[i*3+0] as u32 + pixels[i*3+0] as u32 + pixels[i*3+0] as u32) / 3;
-                    rgba_pixels[i*4+0] = 0xff;
-                    rgba_pixels[i*4+1] = 0xff;
-                    rgba_pixels[i*4+2] = 0xff;
-                    rgba_pixels[i*4+3] = alpha as u8;
+                    // This is what skia does
+                    let alpha :u8 = ((pixels[i*3+0] as u32 +
```
Comment (vvuk) [inline webrender/src/platform/windows/font.rs]: spacing `alpha: u8`.  Though you don't need the ': u8' at all, the whole thing gets cast at the end anyway.

File: webrender/src/platform/windows/font.rs [hunk 7]
```diff
@@ -85,98 +139,88 @@ impl FontContext {
...
 
-        let pixels = analysis.create_alpha_texture(tex_type, bounds);
-        let rgba_pixels = match render_mode.unwrap() {
+    // DWRITE gives us values in RGB. WR expects it in RGBA. Note, CG returns in BGR
+    // TODO: probably should fix that
+    fn convert_to_rgba(&self, pixels: &Vec<u8>, render_mode: FontRenderMode) -> Vec<u8> {
+        match render_mode {
             FontRenderMode::Mono => {
-                let mut rgba_pixels = vec![0; pixels.len() * 4];
+                let mut rgba_pixels = vec![0 as u8; pixels.len() * 4];
```
Comment (vvuk) [inline webrender/src/platform/windows/font.rs]: do you need the `as u8`? Doesn't hurt to be explicit, though if you want to be explicit I'd prefer `rgba_pixels: Vec<u8>`.  Don't feel strongly about this, though.

File: webrender/src/platform/windows/font.rs [hunk 8]
```diff
@@ -185,8 +229,23 @@ impl FontContext {
...
+        let analysis = self.create_glyph_analysis(font_key, size, glyph, render_mode);
+        let texture_type = self.get_dwrite_texture_type(render_mode);
+
+        let bounds = analysis.get_alpha_texture_bounds(texture_type);
+        let width = (bounds.right - bounds.left) as usize;
+        let height = (bounds.bottom - bounds.top) as usize;
+
+        let mut pixels = analysis.create_alpha_texture(texture_type, bounds);
+        self.gdi_gamma_lut.preblend_rgb(&mut pixels, width, height,
+                                    Color::new(color.r, color.g, color.b, color.a));
```
Comment (vvuk) [inline webrender/src/platform/windows/font.rs]: Weird indentation here

File: webrender/src/platform/windows/font.rs (hunk 1 above)
Comment (changm) [inline webrender/src/platform/windows/font.rs]: We don't use the gdi_gamma_lut yet. We will once https://github.com/servo/webrender/pull/770 lands.

**Comment by changm** [comment]
@vvuk Thanks, fixed the other things. r?

**Comment by vvuk** [comment]
Uber nit, but `dwrite_texture_type`, no get_ prefix on any of them. (You're not "getting" anything, it's more akin to a cast than anything else.) But can just fix that in a follow-up whoever touches this next.

r+

**Comment by vvuk** [comment]
Oh also - fix what? in the comment about "we should fix this"

**Comment by changm** [comment]
@vvuk Fixed and removed the get_ prefix. Also, I edited the comment in the push to say what we should fix. Not sure how, but GitHub still says changes requested. Not sure how you flip that to r+.

**Comment by vvuk** [comment]
I restarted the failing build.. timed out after 10 min, are we hitting a cap that needs to be raised?

**Pull request merged**